### PR TITLE
Task frogger game added

### PR DESCRIPTION
**Pull request by BlueLamantine**
### OOP JS (Frogger game)

[Demo](https://bluelamantine.github.io/kottans-nanodegree-arcade-game/) |
[Code base](https://github.com/BlueLamantine/kottans-nanodegree-arcade-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/BlueLamantine/task-js-opp/app.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            if (this.x > STEP) this.x -= STEP;
+        },
+        'right': () => {
+            if (this.x < STEP*4) this.x += STEP;
+        },
+        'default' : () => {
+            return false;
+        }
+    };
+    return (move[key] || move['default'])();
```
Comment (OleksiyRudenko) [inline submissions/BlueLamantine/task-js-opp/app.js:81]: Nice and smart touch :)

File: submissions/BlueLamantine/task-js-opp/app.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        player.setDefault();
+    }
+   // handles collision with the Player
+   allEnemies.forEach(function(enemy) {
+    if ( Math.round(enemy.x / STEP) === Math.round(player.x / STEP) && 
+         Math.round(enemy.y / BLOCK_HEIGHT) === Math.round(player.y / BLOCK_HEIGHT) ) {
+            alert('you lose! =(');
+            player.setDefault();
+    }
+   });
```
Comment (OleksiyRudenko) [inline submissions/BlueLamantine/task-js-opp/app.js:57]: If you had to add a comment this is a clear sign that this part is worth being isolated into a function/method.
Besides `update` does two very diifferent things : check location and check collisions. This is violation of the SIngle Responsibility Principle.

Solution:
Define a dedicated function to check against collisions.
Call it from `update`

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@BlueLamantine a really good job!
A couple of fixes and you're done.

File: submissions/BlueLamantine/task-js-opp/app.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    this.y = y;
+    this.speed = speed;
+    this.sprite = 'images/cat.gif';
+};
+
+let allEnemies =  Array.from(Array(ENEMY_CONF.quantity),
+(val, index) => new Enemy(ENEMY_CONF.startPosX,ENEMY_CONF.startPosY[index],getRandomNumber(MIN_SPEED,MAX_SPEED)));
+
+Enemy.prototype.update = function(dt) {
+    this.x > WIDTH ? this.x = ENEMY_CONF.startPosX : this.x = this.x + this.speed * dt;
```
[the remark this answers is not in the document]
Comment (BlueLamantine) [reply submissions/BlueLamantine/task-js-opp/app.js]: ok, agreed and changed a little, now it is more like an example!

File: submissions/BlueLamantine/task-js-opp/app.js (hunk 2 above)
Comment (BlueLamantine) [reply submissions/BlueLamantine/task-js-opp/app.js]: Thanks for comment, my bad. I fixed it according to your proposed solution!

**Comment by BlueLamantine** [comment]
@OleksiyRudenko Thanks for attention to my work! I followed your recommendations.

File: submissions/BlueLamantine/task-js-opp/app.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        player.setDefault();
+    }
+    this.checkCollision();
+};
+Player.prototype.checkCollision = function () {
+    allEnemies.forEach(function(enemy) {
+        if ( Math.round(enemy.x / STEP) === Math.round(player.x / STEP) && 
+             Math.round(enemy.y / BLOCK_HEIGHT) === Math.round(player.y / BLOCK_HEIGHT) ) {
+                alert('you lose! =(');
+                player.setDefault();
```
Comment (OleksiyRudenko) [inline submissions/BlueLamantine/task-js-opp/app.js:57]: It is against OOP and SOLID principles to refer to global variables from inside the class.

Exclusions are constants and public API.

`allEnemies` and `player` are global variables here.

`allEnemies`: just let `Player` instance legally know about them at a player construction time.

`player`: we do need to refer to a player variable here at all as we are already "inside" the player. Hope it is obvious how to refer the object from inside itself.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@BlueLamantine well done!
Need a fix so that OOP/SOLID principles are followed.

File: submissions/BlueLamantine/task-js-opp/app.js (hunk 4 above)
Comment (BlueLamantine) [reply submissions/BlueLamantine/task-js-opp/app.js]: Ok, I fixed . The function ```checkCollision();``` does not refer to global variables except constants (hope that present global variable ```allEnemies``` as a parameter of constructor ```Player``` is a legal way)

Comment (OleksiyRudenko) [reply submissions/BlueLamantine/task-js-opp/app.js]: It's absolutely legal. This way a entity (character) learns improtant things about surrounding world.

**Review by OleksiyRudenko — APPROVED** [review]
@BlueLamantine great job done!

**Pull request merged**